src/main/java/org/traccar/protocol/Xexun2ProtocolEncoder.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2016 - 2022 Anton Tananaev (anton@traccar.org)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.traccar.protocol;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+import org.traccar.BaseProtocolEncoder;
+import org.traccar.helper.DataConverter;
+import org.traccar.model.Command;
+import org.traccar.Protocol;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+public class Xexun2ProtocolEncoder extends BaseProtocolEncoder {
+
+    public Xexun2ProtocolEncoder(Protocol protocol) {
+        super(protocol);
+    }
+
+    private static ByteBuf encodeFrame(ByteBuf buf) {
+        int bufLength = buf.readableBytes();
+        if (bufLength < 5) {
+            return null;
+        }
+
+        ByteBuf result = Unpooled.buffer();
+
+        result.writeBytes(buf.readBytes(2));
+
+        while (buf.readerIndex() < bufLength - 2) {
+            int b = buf.readUnsignedByte();
+            if (b == 0xfa && buf.isReadable() && buf.getUnsignedByte(buf.readerIndex()) == 0xaf) {
+                buf.readUnsignedByte();
+                result.writeByte(0xfb);
+                result.writeByte(0xbf);
+                result.writeByte(0x01);
+            } else if (b == 0xfb && buf.isReadable() && buf.getUnsignedByte(buf.readerIndex()) == 0xbf) {
+                buf.readUnsignedByte();
+                result.writeByte(0xfb);
+                result.writeByte(0xbf);
+                result.writeByte(0x02);
+            } else {
+                result.writeByte(b);
+            }
+        }
+        result.writeBytes(buf.readBytes(2));
+
+        return result;
+    }
+
+    private static int checksum(byte[] data)
```

Comment:
Are you sure it's not one of the standard checksums we already support?

Reply:
I've tried all the CRC16 algorithms and none generate the same checksum.

Reply:
Can you please include the documentation.

Reply:
https://www.traccar.org/protocol/5233-xexun2/XEXUN%20Devices%20and%20Server%20Data%20Protocol%20for%20GPS%20tracker%2020210520.pdf

1.4.4 ECC on page 3&4 - no description on algorithm, just java & C code provided.

Reply:
Looks like it's a standard TCP/UDP checksum. We should add it to our standard checksum class.

http://minirighi.sourceforge.net/html/udp_8c-source.html

Reply:
Also you should use buffer, not array.